### DISCUSSION TITLE
WatsonTcp as a Windows Service

### DISCUSSION THREAD

**Discussion by Tithras** [jchristn/WatsonTcp#193] [General]
Good morning, I have currently got a console server which accepts multiple connections and outputs to the console when connections / disconnects happen etc. Fairly basic and works with no issues, the only thing I did extra was allow 1 client to send to all clients as that was 1 of my requirements.

My question is that I have attempted to migrated it to a windows service rather than console, with the debugging and logging output going to a log file however have run into issues. The service starts, log file states accepting clients and i can then connect a client however if I disconnect that client and reconnect it i find the service has stopped. I must assume that I am getting some sort of exception as part of that connection / disconnection process but I am at a loss to find what is causing it. Has anyone been able to get this to run as a windows service successfully and if so would you be able to show how you got it working? Thanks.

**Comment by Tithras** [comment]
I should also state that disconnection and reconnection works fine on the console only app!

**Comment by Tithras** [comment]
Another addition, i have removed any of the StreamReceived code outside of the function exists with nothing in it. Service still stops, defiantly appears to be related to the connect / disconnect code although no idea where.

**Comment by jchristn** [comment]
Hi @Tithras I will take a look at this ASAP!  Thanks for letting me know.

**Reply by Tithras** [reply]
Thanks, it would be good to know that it isnt something I have done wrong :)

**Reply by Tithras** [reply]
This is my service OnStart() method:

protected override void OnStart(string[] args)
{
     // Setup watson client defaults
     tcpServer = new WatsonTcpServer(null, serverPort);
     tcpServer.Events.ClientConnected += MyClientConnected;
     tcpServer.Events.ClientDisconnected += MyClientDisconnected;
     tcpServer.Events.StreamReceived += MyStreamReceived;
     tcpServer.Settings.DebugMessages = MyDebugMessages;
     tcpServer.Settings.Logger = MyLogger;
     tcpServer.Events.ExceptionEncountered += MyExceptionEncountered;

     // Start OEMS server
     tcpServer.Start();

     // Log that server started
     MyLogger(Severity.Info, "Server started.");
}

The ClientConnected and ClientDisconnecd have nothing but a Logger call.
The StreamReceived has a ton of stuff in BUT if I remove it all then I still get the same issue.

**Reply by Tithras** [reply]
I have to laugh, it was an issue at my end in the end. I was writing to a log file but running into io exception errors which even though it was in a try/catch (the log file writing) didnt throw the exception it just died. Please ignore this entire post, sorry for the trouble!

**Reply by jchristn** [reply]
Hi @Tithras well, I'm glad to hear you got it working!  I had literally just built a project to test this (it was successful).  I'll post a link to it momentarily :)

**Reply by jchristn** [reply]
https://github.com/jchristn/WatsonWindowsService

**Comment by Tithras** [comment]
Wow thanks :)

Managed to resolve my issue with a threadsafe logging method so it works fine now but I will have a look through your example to confirm I haven't missed anything else :)

**Reply by jchristn** [reply]
Here's how I did it:
```
private string _Logfile = "./watson.log";
private readonly object _Lock = new object();
```
and then
```
private void Log(string msg)
{
  if (String.IsNullOrEmpty(msg)) return;
  if (!msg.EndsWith(Environment.NewLine)) msg += Environment.NewLine;
  lock (_Lock)
  {
    File.AppendAllText(_Logfile, msg);
  }
}
```
and usage such as:
```
private void ClientDisconnected(object sender, DisconnectionEventArgs e)
{
  Log("Client " + e.IpPort + " disconnected");
}
```

**Comment by Tithras** [comment]
Yup, pretty much the same :)

Again, thanks for taking the time to look even though it turned out to be my naff coding !!!

The joys of being a hobby coder trying to write production code ;)
